### DISCUSSION TITLE
env.AddBuildMiddleware not working ?

### DISCUSSION THREAD

**Discussion by gneiss15** [arendst/Tasmota#19967] [Support request]
Because one of my files needs (at least) the C++17 standard I am looking for a way to switch to this standard.
I tried several aproches.

- The (only) one that works is to change "-std=gnu++11" inside ".platformio/packages/framework-arduinoespressif32/tools/platformio-build-esp32.py" to "-std=gnu++17".
Using "-std=c++17" instead would produce lots of errors in different files of the project.
IMHO this aproach isn't the one that should be used, because it changes something outside of Tasmota.
So it will not be part of the repository :-(
- All tries to use BUILD_UNFLAGS doesn't work (don't know why)
- Before I found the above working solution, I had tried to add a script that changes the flag only for the file that needs it.
I wrote a (pos_extra_)script that uses env.AddBuildMiddleware to change the environment for this file.
Inside this script I althought create a dump of env & projenv. This part is working.
But even the simplesd AddBuildMiddleware does not run.

The usage of  AddBuildMiddleware is documented here:
https://docs.platformio.org/en/latest/scripting/middlewares.html
and I found it from here:
https://community.platformio.org/t/apply-build-flag-for-a-specific-file-only/32024

Here is my code snipped:
```
Import("env", "projenv")

# Dump global construction environment (for debug purpose) <- That works
with open( '/home/gneiss/VSCode-linux-x64/data/Tasmota/build_output/build_env.log', 'w' ) as f:
  print( env.Dump(), file = f )

# Dump project construction environment (for debug purpose) <- That works, too
with open( '/home/gneiss/VSCode-linux-x64/data/Tasmota/build_output/build_projenv.log', 'w' ) as f:
  print( projenv.Dump(), file = f )

# this function is never called ???
def ShowFiles( env, node ):
  with open( '/home/gneiss/VSCode-linux-x64/data/Tasmota/build_output/build_files.log', 'a' ) as f:
    print( node.name, file = f )

env.AddBuildMiddleware( ShowFiles )
```

**Comment by barbudor** [comment]
Good luck because frankly you won't get help
Tasmota compile as it is.
If you want to integrate something in Tasmota change it to fit tasmota constraint

**Reply by gneiss15** [reply]
Yeah, and even that the project build with "-std=gnu++17" inside "..platformio-build-esp32.py",it will crash imediate after upload
:-((
But that has (IMHO) nothing to do with "AddBuildMiddleware" not working..

**Reply by Staars** [reply]
Just use/modify an environment that uses Arduino 3. This will use a newer compiler toolchain and your particular problem should magically disappear.
Besides you are indeed on your own, when you fiddle with the code base, but this is the way.

**Comment by gneiss15** [comment]
OK, the solution is simpler than expected :-)
I now just use (in platformio.ini):
```
[esp_defaults]
...
build_unflags               = ...
                              -std=gnu++11
build_flags                 = ...
                              -std=gnu++17
```
And inside the file that need's C++17, I use:

```
#if ! defined( __cplusplus ) | __cplusplus < 201703L
  #error This file must be compiled in C++ Mode with (at least) C++17 standard (set -std=gnu++17)
#endif
```

It seams, that a lot portions of Tasmota needs the GNU C++ extension, that is the reason why -std=c++17 is not working..
The previrious reported crash when using C++17 has another reason.

**But it is still anoing and unanswered, that/why "AddBuildMiddleware" is not working..**

**Comment by Jason2866** [comment]
General Platformio and C++ questions are not discussed here.
